relay_messages.go

```diff
@@ -134,22 +141,38 @@ func newLazyCallReq(f *Frame) (*lazyCallReq, error) {
 	}
 
 	// csumtype:1 (csum:4){0,1} arg1~2 arg2~2 arg3~2
-	checkSumType := ChecksumType(rbuf.ReadSingleByte())
-	rbuf.ReadBytes(checkSumType.ChecksumSize())
+	cr.checksumTypeOffset = rbufOffset()
+	cr.checksumType = ChecksumType(rbuf.ReadSingleByte())
+	rbuf.ReadBytes(cr.checksumType.ChecksumSize())
 
 	// arg1~2
 	arg1Len := int(rbuf.ReadUint16())
 	cr.method = rbuf.ReadBytes(arg1Len)
 
 	// arg2~2
-	arg2Len := int(rbuf.ReadUint16())
-	cr.arg2StartOffset = int(headerStartIndex) + rbuf.BytesRead()
-	// TODO(echung): deprecate the arg2 offsets and store just the arg2 slice instead
-	// Read past arg2
-	rbuf.ReadBytes(arg2Len)
+	arg2Len := rbuf.ReadUint16()
+	cr.arg2StartOffset = rbufOffset()
 	cr.arg2EndOffset = cr.arg2StartOffset + arg2Len
+
 	// arg2 is fragmented if we don't see arg3 in this frame.
-	cr.isArg2Fragmented = rbuf.BytesRemaining() == 0 && cr.HasMoreFragments()
+	if uint16(rbuf.BytesRemaining()) <= arg2Len {
+		rbuf.ReadBytes(rbuf.BytesRemaining())
+		cr.isArg2Fragmented = cr.HasMoreFragments()
+	} else {
+		rbuf.ReadBytes(int(arg2Len))
+	}
+
+	if !cr.isArg2Fragmented {
+		// arg3~2
+		arg3Len := rbuf.ReadUint16()
+		cr.arg3StartOffset = rbufOffset()
+
+		if uint16(rbuf.BytesRemaining()) < arg3Len {
```

Comment:
as above, don't think this if is necessary. in this case, `BytesRemaining == arg3Len`, otherwise we have a protocol error.

given that, I don't think we need `arg3EndOffset`. It always ends at the frame boundary since there's nothing else after it in the protocol

I think we should do something like:
```go
if arg3Len != rbuf.BytesRemaining() {
  // return error about the mismatched arg3
}

```